ml-agents/mlagents/trainers/agent_processor.py

```diff
@@ -68,31 +65,28 @@ def add_experiences(
                 "Policy/Learning Rate", take_action_outputs["learning_rate"]
             )
 
-        for agent_id in curr_info.agents:
-            self.last_brain_info[agent_id] = curr_info
+        for agent_id in previous_action.agents:
             self.last_take_action_outputs[agent_id] = take_action_outputs
 
-        # Store the environment reward
-        tmp_environment_reward = next_info.rewards
-
-        for next_idx, agent_id in enumerate(next_info.agents):
-            stored_info = self.last_brain_info.get(agent_id, None)
-            if stored_info is not None:
-                stored_take_action_outputs = self.last_take_action_outputs[agent_id]
-                idx = stored_info.agents.index(agent_id)
-                obs = []
-                if not stored_info.local_done[idx]:
-                    for i, _ in enumerate(stored_info.visual_observations):
-                        obs.append(stored_info.visual_observations[i][idx])
-                    if self.policy.use_vec_obs:
-                        obs.append(stored_info.vector_observations[idx])
+        for agent_id in batched_step_result.agent_id:
+            curr_agent_step = batched_step_result.get_agent_step_result(agent_id)
+            stored_step = self.last_step_result.get(agent_id, None)
+            stored_take_action_outputs = self.last_take_action_outputs.get(
+                agent_id, None
+            )
+            if stored_step is not None and stored_take_action_outputs is not None:
+                stored_agent_step = stored_step.get_agent_step_result(agent_id)
+
+                idx = stored_step.agent_id.index(agent_id)
```

Comment:
BatchedStepResult._agent_id_to_index is a faster way to get this. It's currently lazily built in contains_agent(), but we could move it to a property and build it there instead.

Reply:
I do like using that dict instead since it likely has already been built the previous timestep. I didn't change it in this PR since it would be a change in the public-facing Python LL-API. What if we do a public agent_id_to_index property and move the creation of the dict there, and then just access the property in contains_agent()?

It seems like the core problem, though, is the fact that we need this index at all. Andrew and I were discussing a change that would stop us from passing the action back from the environment manager, and rather just grabbing it from the Policy that took the action. So hopefully anything we do here is temporary.

Reply:
Made agent_id_to_index a property